promql/lex.go

```diff
@@ -0,0 +1,693 @@
+
+package promql
+
+import (
+	"fmt"
+	"reflect"
+	"strings"
+	"unicode"
+	"unicode/utf8"
+)
+
+// item represents a token or text string returned from the scanner.
+type item struct {
+	typ itemType // The type of this item.
+	pos Pos      // The starting position, in bytes, of this item in the input string.
+	val string   // The value of this item.
+}
+
+// String returns a descriptive string for the item.
+func (i item) String() string {
+	switch {
+	case i.typ == itemEOF:
+		return "EOF"
+	case i.typ == itemError:
+		return i.val
+	case i.typ.isKeyword():
+		return fmt.Sprintf("<%s>", i.val)
+	case i.typ.isOperator():
+		return fmt.Sprintf("<op:%s>", i.val)
+	case i.typ.isAggregator():
+		return fmt.Sprintf("<aggr:%s>", i.val)
+	case len(i.val) > 10:
+		return fmt.Sprintf("%.10q...", i.val)
+	}
+	return fmt.Sprintf("%q", i.val)
+}
+
+// isOperator returns true if the item corresponds to a logical or arithmetic operator.
+// Returns false otherwise.
+func (i itemType) isOperator() bool { return i > operators_start && i < operators_end }
+
+// isAggregator returns true if the item belongs to the aggregator functions.
+// Returns false otherwise
+func (i itemType) isAggregator() bool { return i > aggregators_start && i < aggregators_end }
+
+// isKeyword returns true if the item corresponds to a keyword.
+// Returns false otherwise.
+func (i itemType) isKeyword() bool { return i > keywords_start && i < keywords_end }
+
+// Constants for operator precedence in expressions.
+const (
+	LowestPrec  = 0 // non-operators
+	HighestPrec = 7
+)
+
+// Precedence returns the operator precedence of the binary
+// operator op. If op is not a binary operator, the result
+// is LowestPrec.
+func (i itemType) precedence() int {
+	switch i {
+	case itemLOR:
+		return 1
+	case itemLAND:
+		return 2
+	case itemEQL, itemNEQ, itemLTE, itemLSS, itemGTE, itemGTR:
+		return 3
+	case itemADD, itemSUB:
+		return 4
+	case itemMUL, itemQUO, itemREM:
+		return 5
+	}
+	return LowestPrec
+}
+
+type itemType int
+
+const (
+	itemError itemType = iota // Error occurred, value is error message
+	itemEOF
+	itemComment
+	itemIdentifier
+	itemMetricIdentifier
+	itemLeftParen
+	itemRightParen
+	itemLeftBrace
+	itemRightBrace
+	itemLeftBracket
+	itemRightBracket
+	itemComma
+	itemAssign
+	itemSemicolon
+	itemString
+	itemNumber
+	itemDuration
+
+	operators_start
+	// operators
+	itemSUB
+	itemADD
+	itemMUL
+	itemREM
+	itemQUO
+	itemLAND
+	itemLOR
+	itemEQL
+	itemNEQ
+	itemLTE
+	itemLSS
+	itemGTE
+	itemGTR
+	itemEQLRegex
+	itemNEQRegex
+	operators_end
+
+	aggregators_start
+	// aggregators
+	itemAvg
+	itemCount
+	itemSum
+	itemMin
+	itemMax
+	itemStddev
+	itemStdvar
+	aggregators_end
+
+	keywords_start
+	// keywords
+	itemAlert
+	itemIf
+	itemFor
+	itemWith
+	itemSummary
+	itemDescription
+	itemKeepingExtra
+	itemOffset
+	itemBy
+	itemOn
+	itemGroupLeft
+	itemGroupRight
+	keywords_end
+)
+
+var key = map[string]itemType{
+	// operators
+	"AND": itemLAND,
```

Comment:
Should we switch the "canonical" form of the keywords to lower-case? For most of them we already practically only use the lower-cased version, except in alerting rules.